File: 250812/아름다운 수/beautiful-number.cpp
#include <iostream>
#include <vector>

using namespace std;

int n, total = 0;
vector<int> answer;

// 아름다운 수 인지 체크하는 함수
int isBeautiful(){
    int idx = 0;
    
    while (idx < n){
        int num = answer[idx];

        for (int i = idx+1; i < idx+num; i++){
            if (i >= n || answer[i] != num) return 0;
        }
        
        idx += num;
    }

    return 1;
}

// 모든 경우의 수를 구하는 함수
void Choose(int curr_num){
    if (curr_num == n+1){
        total += isBeautiful();
        return ;
    }

    for (int i = 1; i <= 4; i++){
        answer.push_back(i);
        Choose(curr_num+1);
        answer.pop_back();
    }
}

int main(){
    cin >> n;
    Choose(1);

    cout << total;
    return 0;
}

File: 251016/아름다운 수/beautiful-number.py
n = int(input())
answer = 0
arr = []

def check():
    global answer
    i = 0
    while i < n:
        num = arr[i]
        if num+i > n:  # arr의 범위를 벗어남
            return

        for j in range(i + 1, i + num):
            if num != arr[j]:
                return
        i += num

    answer += 1
    return

def choose(curr_num):
    # 종료 조건
    if curr_num == n + 1:
        check()
        return

    # 1~4 반복 시작
    for i in range(1, 5):
        arr.append(i)
        choose(curr_num + 1)
        arr.pop()

    return


choose(1)

print(answer)